op-node/cmd/span_batch_tester/convert/convert.go

```diff
@@ -0,0 +1,158 @@
+package convert
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"path"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/cmd/batch_decoder/reassemble"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+type Config struct {
+	InDirectory      string
+	OutDirectory     string
+	GenesisTimestamp uint64
+}
+
+type SpanBatchWithMetadata struct {
+	BatchV2     derive.BatchV2
+	BatchV2Hash []byte
+	L1StartNum  uint64
+	L1EndNum    uint64
+	L2StartNum  uint64
+	L2EndNum    uint64
+}
+
+func LoadChannels(dir string) []reassemble.ChannelWithMetadata {
+	files, err := os.ReadDir(dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var out []reassemble.ChannelWithMetadata
+	for _, file := range files {
+		f := path.Join(dir, file.Name())
+		chm := LoadChannelFile(f)
+		out = append(out, chm)
+	}
+	return out
+}
+
+func LoadChannelFile(file string) reassemble.ChannelWithMetadata {
+	f, err := os.Open(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	dec := json.NewDecoder(f)
+	var chm reassemble.ChannelWithMetadata
+	if err := dec.Decode(&chm); err != nil {
+		log.Fatalf("Failed to decode %v. Err: %v\n", file, err)
+	}
+	return chm
+}
+
+// GetL2StartNum returns L2 block start number per channel
+func GetL2StartNum(client *ethclient.Client, chm reassemble.ChannelWithMetadata) (uint64, error) {
+	if len(chm.Batches) == 0 {
+		return 0, errors.New("channel is empty")
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	startParent, err := client.BlockByHash(ctx, chm.Batches[0].ParentHash)
+	if err != nil {
+		return 0, err
+	}
+	startNum := startParent.NumberU64() + 1
+	return startNum, nil
+}
+
+// GetOriginChangedBit returns bit indicating L1Origin has changed for given l2 block number compared to its parent
+func GetOriginChangedBit(client *ethclient.Client, l2BlockNum uint64, l1OriginHash common.Hash) (uint, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	parentBlock, err := client.BlockByNumber(ctx, new(big.Int).SetUint64(l2BlockNum-1))
```

Comment:
parent block could be cached in `GetL2StartNum`

Reply:
Patched at https://github.com/testinprod-io/optimism/pull/4/commits/fd27e7612c0792c90be00dc8c437848b926334d1